contrib/src/main/scala/akka/stream/contrib/FocusedFlow.scala

```diff
@@ -0,0 +1,241 @@
+   graph: Graph[IncompleteFocusedSourceShape[(Tag, In), In, Out], Mat]
+  ) extends FlowOpsMat[Out, Mat] {
+  import GraphDSL.Implicits._
+
+  override type Repr[+O] = FocusedSource[Tag, In, O, Mat@uncheckedVariance]
+  override type ReprMat[+O, +M] = FocusedSource[Tag, In, O, M]
+  override type Closed = RunnableGraph[Mat@uncheckedVariance]
+  override type ClosedMat[+M] = RunnableGraph[M]
+
+  def unfocus: Source[(Tag, Out), Mat] = Source.fromGraph(
+    GraphDSL.create(graph) {
+      implicit builder =>
+      g =>
+        val tagger = builder.add(new FlowTagger[In, Out, Tag])
+
+        g.srcIn ~> tagger.in1
+        tagger.out2 ~> g.in
+        g.out ~> tagger.in2
+
+        SourceShape(tagger.out1)
+    })
+
+  override def via[T, Mat2](flow: Graph[FlowShape[Out, T], Mat2]): FocusedSource[Tag, In, T, Mat] =
+    viaMat(flow)(Keep.left)
+  override def viaMat[T, Mat2, Mat3](flow: Graph[FlowShape[Out, T], Mat2])(combine: (Mat, Mat2) => Mat3): FocusedSource[Tag, In, T, Mat3] =
+    new FocusedSource(GraphDSL.create(graph, flow)(combine) {
+      implicit builder =>
+        (g, f) =>
+          g.out ~> f
+          IncompleteFocusedSourceShape(g.srcIn, g.in, f.out)
+    })
+
+  // connecting the focused flow to a sink will discard the tag
+  override def to[Mat2](sink: Graph[SinkShape[Out], Mat2]): RunnableGraph[Mat] =
+  toMat(sink)(Keep.left)
+  override def toMat[Mat2, Mat3](sink: Graph[SinkShape[Out], Mat2])(combine: (Mat, Mat2) => Mat3): RunnableGraph[Mat3] =
+    RunnableGraph.fromGraph(
+      GraphDSL.create(graph, sink)(combine) {
+        implicit builder =>
+          (g, s) =>
+            val mapper = builder.add(new akka.stream.impl.fusing.Map[(_, In), In](_._2))
+            g.srcIn ~> mapper ~> g.in
+            g.out ~> s.in
+            ClosedShape
+      })
+
+  override def mapMaterializedValue[Mat2](f: (Mat) => Mat2): FocusedSource[Tag, In, Out, Mat2] =
+    viaMat(Flow[Out]) { case (m, notused) => f(m) }
+
+  override def withAttributes(attr: Attributes): FocusedSource[Tag, In, Out, Mat] =
+    new FocusedSource(graph.withAttributes(attr))
+
+  override def addAttributes(attr: Attributes): FocusedSource[Tag, In, Out, Mat] =
+    new FocusedSource(graph.addAttributes(attr))
+
+  override def named(name: String): FocusedSource[Tag, In, Out, Mat] =
+    new FocusedSource(graph.named(name))
+
+  override def async: FocusedSource[Tag, In, Out, Mat] =
+    new FocusedSource(graph.async)
+}
+
+private class FlowTagger[In, Out, E]
+  extends GraphStage[BidiShape[(E, In), (E, Out), Out, In]] {
+  val inTup = Inlet[(E, In)]("in-tuple")
+  val outTup = Outlet[(E, Out)]("out-tuple")
+  val toFlow = Outlet[In]("to-flow")
+  val fromFlow = Inlet[Out]("from-flow")
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    var curr: E = _
+    setHandler(inTup, new InHandler {
+      override def onPush(): Unit = {
+        val (cur, in) = grab(inTup)
+        curr = cur
```

Comment:
Actually going with raw, sbt-boilerplate generated ones would be enough it seems,
and it also seems we do need a custom type, SubFlow won't do hm.

General idea would be:

```
  // need to generate for 22-arity
  implicit final class Focus2OnSource[+T1, +T2, +M](s: Source[(T1, T2), M]) {
  import TupleGet._
    def focus[T, TT1, TT2]( // fix variance
      get: Tuple2[T1, T2] => T, 
      set: Tuple2[T1, T2] => T => Tuple2[TT1, TT2]): impl.SubFlowImpl[T, T, M, scaladsl.Source[Out, M]#Repr, scaladsl.RunnableGraph[M]] = ???
  }

  Source.single("" -> 12)
    .focus(get = t => t._1, 
           set = t => v => t.copy(_1 = v))
```

And instead of the SubFlow we'd need a custom type there it seems though.
Would you be able to give it a shot towards n-arity? Then it would be really awesome and useful I think